File: lib/models/movies.dart
class Movies {
  final int id;
  final String title;
  final String originalTitle;
  final String posterPath;
  final String backdropPath;
  final String? overview;
  final String? localPath;
  final String releaseDate;
  final double voteAverage;
  final int voteCount;
  final String originalLanguage;
  final List<int> genreIds;
  final List<String> castNames;
  final List<String> crewNames;

  Movies({
    required this.id,
    required this.title,
    this.originalTitle = '',
    required this.posterPath,
    this.backdropPath = '',
    this.overview,
    this.localPath,
    this.releaseDate = '',
    this.voteAverage = 0,
    this.voteCount = 0,
    this.originalLanguage = '',
    this.genreIds = const [],
    this.castNames = const [],
    this.crewNames = const [],
  });

  String get year =>
      releaseDate.isNotEmpty ? releaseDate.substring(0, 4) : '';
  String get rating => voteAverage.toStringAsFixed(1);

  factory Movies.fromJson(Map<String, dynamic> json) {
    return Movies(
      id: json['id'] ?? 0,
      title: json['title'] ?? '',
      originalTitle: json['original_title'] ?? '',
      posterPath: json['poster_path'] ?? '',
      backdropPath: json['backdrop_path'] ?? '',
      overview: json['overview'] ?? '',
      localPath: json['localPath'] ?? json['local_path'] ?? '',
      releaseDate: json['release_date'] ?? '',
      voteAverage: (json['vote_average'] ?? 0).toDouble(),
      voteCount: json['vote_count'] ?? 0,
      originalLanguage: json['original_language'] ?? '',
      genreIds: List<int>.from(json['genre_ids'] ?? []),
      castNames: List<String>.from(json['cast_names'] ?? []),
      crewNames: List<String>.from(json['crew_names'] ?? []),
    );
  }
}

File: server/routes/movies.js
const express = require('express');
const router = express.Router();
const fs = require('fs');
const path = require('path');
const fetch = (...args) => import('node-fetch').then(({default: f}) => f(...args));

const FILMS_DIR = process.env.FILMS_DIR;
const TMDB_API_KEY = process.env.TMDB_API_KEY;
const TMDB_BASE_URL = 'https://api.themoviedb.org/3';

const { getCache, setCache } = require('../cache');
let lastScan = null;

function parseFilename(filename) {
  const noExt = path.basename(filename, path.extname(filename));
  const match = noExt.match(/^(.+?)_(\d{4})$/);
  if (match) {
    return { title: match[1].replace(/_/g, ' ').trim(), year: match[2] };
  }
  return { title: noExt.replace(/_/g, ' ').trim(), year: null };
}

async function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function searchTmdb(title, year) {
  try {
    const query = encodeURIComponent(title);
    const yearParam = year ? `&year=${year}` : '';
    const url = `${TMDB_BASE_URL}/search/movie?api_key=${TMDB_API_KEY}&query=${query}${yearParam}&language=it-IT`;
    const res = await fetch(url);
    const data = await res.json();
    if (data.results && data.results.length > 0) return data.results[0];
  } catch (e) {
    console.error(`Errore TMDB per "${title}":`, e.message);
  }
  return null;
}

async function fetchCredits(movieId) {
  try {
    const url = `${TMDB_BASE_URL}/movie/${movieId}/credits?api_key=${TMDB_API_KEY}&language=it-IT`;
    const res = await fetch(url);
    const data = await res.json();
    const castNames = (data.cast || []).slice(0, 10).map(a => a.name.toLowerCase());
    const crewNames = (data.crew || [])
      .filter(c => c.job === 'Director')
      .map(d => d.name.toLowerCase());
    return { castNames, crewNames };
  } catch (e) {
    console.error(`Errore credits per movie ${movieId}:`, e.message);
    return { castNames: [], crewNames: [] };
  }
}

async function scanFilms() {
  console.log('Scansione cartella:', FILMS_DIR);
  const files = fs.readdirSync(FILMS_DIR).filter(f =>
    ['.mp4', '.mkv', '.avi', '.mov'].includes(path.extname(f).toLowerCase())
  );
  console.log(`Trovati ${files.length} file video`);

  const movies = [];

  for (const file of files) {
    const { title, year } = parseFilename(file);
    //console.log(`Cerco su TMDB: "${title}" (${year ?? 'anno sconosciuto'})`);
    const tmdb = await searchTmdb(title, year);

    let castNames = [];
    let crewNames = [];

    if (tmdb?.id) {
      await sleep(300);
      const credits = await fetchCredits(tmdb.id);
      castNames = credits.castNames;
      crewNames = credits.crewNames;
    }

    movies.push({
      id: tmdb?.id ?? Math.floor(Math.random() * 100000),
      title: tmdb?.title ?? title,
      original_title: tmdb?.original_title ?? '',
      poster_path: tmdb?.poster_path ?? '',
      backdrop_path: tmdb?.backdrop_path ?? '',
      overview: tmdb?.overview ?? '',
      release_date: tmdb?.release_date ?? year ?? '',
      vote_average: tmdb?.vote_average ?? 0,
      vote_count: tmdb?.vote_count ?? 0,
      original_language: tmdb?.original_language ?? '',
      genre_ids: tmdb?.genre_ids ?? [],
      popularity: tmdb?.popularity ?? 0,
      cast_names: castNames,
      crew_names: crewNames,
      local_path: path.join(FILMS_DIR, file),
    });
  }

  setCache(movies);
  lastScan = new Date();
  console.log(`Scansione completata: ${movies.length}/${files.length} film trovati`);
  return movies;
}

scanFilms();

router.get('/', async (req, res) => {
  const moviesCache = getCache();
  if (moviesCache.length === 0) await scanFilms();
  res.json(getCache());
});

router.get('/scan', async (req, res) => {
  const movies = await scanFilms();
  res.json({ message: `Scansione completata: ${movies.length} film trovati`, movies });
});

router.get('/:id', (req, res) => {
  const moviesCache = getCache();
  const movie = moviesCache.find(m => m.id === parseInt(req.params.id));
  if (!movie) return res.status(404).json({ error: 'Film non trovato' });
  res.json(movie);
});

module.exports = router;